Add backupConfiguration to Analysis services request/response properties

ziyuanms opened the pull request:
## Description

This change is to support backup when create, update or get server
details. backup configuration has 3 properties, storage account and blob
container are required for all operations. Access key is required for
create and update operations.

## PR information

- [x] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, see this page.
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the api-version in the path should match the api-version in the spec).

## Quality of Swagger

- [x] I have read the contribution guidelines.
- [ ] My spec meets the review criteria:
  - [x] The spec conforms to the Swagger 2.0 specification.
  - [x] Validation errors from the Linter extension for VS Code have all been fixed for this spec. (Note: for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [ ] The spec follows the patterns described in the Swagger good patterns document unless the service API makes this impossible.

ziyuanms committed:
Add backupConfiguration to Analysis services request/response properties

This change is to support backup when create, update or get server
details. backup configuration has 3 properties, storage account and blob
container are required for all operations. Access key is required for
create and update operations.

azurecla commented:
Hi __@ziyuanms__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (ziwan). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

ziyuanms commented:
Hi @jhendrixMSFT , I see related changes in azure-sdk-for-net has been merge (https://github.com/Azure/azure-sdk-for-net/pull/2831). can you review and merge this request as well? Thanks.

amarzavery commented:
@johanste, @ravbhatnagar - Calling this out. This will be a breaking change. Not sure how the API review board approved this? How can an RP add extra properties in the payload without bumping the api-version?

johanste commented:
@amarzavery, if this was an instance of correcting a previously incorrect swagger document, then it wouldn't have gone through the API review board. If not, the API version should indeed have been incremented.

ziyuanms commented:
@amarzavery  @johanste  

1. This is backward compatible, the new property BackupConfiguration can be null, without affecting functionality
2. Analysis Services APIs are not released yet, it is still in preview stage for development. 

So I think it's safe to not bump the version.

johanste commented:
There may be unintended consequences for clients using the management libraries, primarily for strongly typed languages. 

The typical example is a client doing an update of an existing resource using a previously generated client library using the GET of the existing resource, modify some properties and then call the Create/PUT operation. If the existing resource had a backup configuration, it would have been removed as part of the update since the earlier version of the management library would have dropped the unknown property returned in the GET call.

In this specific case, the likelihood of this happening is reduced due to the fact that there is a PATCH/update operation supported, which would be the preferred/recommended way to perform the update.

So, in general, from a client library perspective, it is *not* safe to not bump the API version for changes like this. If there would have been a client library that had been released as a stable version, it would very likely cause problems. As it stands today, I don't believe that we do have a non-preview library, so we are likely good on that front as well.

jhendrixMSFT commented:
@amarzavery @johanste Are there any more concerns or are we ok to merge this?

johanste commented:
There are some concerns from the API design perspective (specifically, how storage accounts are referenced).

johanste commented on arm-analysisservices/2016-05-16/swagger/analysisservices.json:
```diff
@@ -525,6 +528,28 @@
+      "properties": {
+        "storageAccount": {
+          "type": "string",
+          "description": "The name of storage account for backup configuration. It is in the format of 'Resource Group Name/Storage Account Name'"
```
This description sounds incorrect based on ARM requirements on referencing resources by IDs rather than custom formats.

ziyuanms replied:
This is how our ARM request/response is working for database backup. For instance, for PATCH operation, the payload is like (copied from armguiclient):

{"properties":{"asAdministrators":{"members":["{admin1}","{admin2}"]},"backupConfiguration":{"storageAccount":"{sa}","blobContainer":"bc","accessKey":"ak"}},"tags":{"{key}":"{value}","{key1}":"{value1}"}}

The implementation will validate storage account, blob container and access key. I think this API change is inline with our ARM operations.

ziyuanms replied:
@johanste I synced with the one who implemented this backup functionality, we have a plan to change to storage account resource id in the near future. We will come back and change here then. For now it is inline with what we have for ARM.

johanste replied:
What does "in the near future" mean in this context? Can SDKs be generated (and shipped) based on the current behavior, and the API version will be updated for the planned changed (correct) behavior? In general, we want to avoid merging swagger specs that we know will change (unless the change includes an API version bump) into master since it makes the SDK release pipeline very difficult to manage.

ziyuanms replied:
@johanste  Understand. We already met with Azure API review board and got exception on this change. We agreed that we will not bump the version before GA.

johanste replied:
@ziyuanms, to be clear, the question is if you intend to *change the behavior* of the API (use full; resource IDs) before you GA. If not, I'm fine with merging. If you are, I'd like to see the description updated before we merge.

ziyuanms replied:
@johanste Just synced with the team, we do have a plan to change the behavior to use full resource id for storage account. So I updated the description accordingly. Thanks.

ziyuanms commented:
@johanste @jhendrixMSFT  Can we merge this change?  Thanks.

ziyuanms committed:
Update the description for storageAccount field of BackupConfiguration

This change is to update the description for storageAccount field of
BackupConfiguration. Before GA, storageAccount will be updated to use
full storage acocunt resource id.

jhendrixMSFT commented:
@amarzavery Any other feedback?

ziyuanms commented:
This is the PR of azure-skd-for-net related to the storage account comment change: https://github.com/Azure/azure-sdk-for-net/pull/2867

@jhendrixMSFT @johanste @amarzavery are we good to merge these changes? Thanks.

amarzavery commented:
LGTM

AutorestCI commented:
No modification for Python

AutorestCI commented:
[![Build Status](https://travis-ci.org/AutorestCI/azure-sdk-for-ruby.svg?branch=RestAPI-PR946)](https://travis-ci.org/AutorestCI/azure-sdk-for-ruby) https://github.com/Azure/azure-sdk-for-ruby/pull/703

AutorestCI commented:
No modification for NodeJS
